Prefer size_t over int where possible

AZero13 opened the pull request:
This avoids needless truncation, which takes up multiple more instructions on some architectures, especially older ones.

enaess commented:
These fixes look mostly good. Though, some cases the refactoring changes the structure ...

enaess commented on pppd/auth.c:
```diff
@@ -2154,22 +2155,24 @@ int
+    do {
 	/* trailing '*' wildcard */
 	l = strlen(wp->word);
+	if (l == 0)
```
Changing from do to while, and checking the length ... should be fine...

enaess commented on pppd/plugins/pppoatm/atm.h:
```diff
@@ -103,6 +103,6 @@ int sap2text(char *buffer,int length,const struct atm_sap *sap,int flags);
 
 int __t2q_get_rate(const char **text,int up);
-int __atmlib_fetch(const char **pos,...); /* internal use only */
+size_t __atmlib_fetch(const char **pos,...); /* internal use only */
```
Should be fine, but a possibly oops here: you can install libatm-dev (on debian) and have pppd's plugin compile against that instead of using the code-drop from the library here.

Is this fix also in the upstream version of libatm-dev?

AtariDreams replied:
I was unaware of this.

enaess commented on pppd/tls.c:
```diff
@@ -186,12 +186,13 @@ static int tls_verify_callback(int ok, X509_STORE_CTX *ctx)
-            if (off > 0) {
-                ptr2 = cn_str + off;
-            }
+
```
Did you test this change? 

As am scrolling through, reviewing this to ensure the functionality have not changed is a bit hard. Though the change make the code look cleaner, does it still work?

AtariDreams replied:
Yes, absolutely.

paulusmack commented:
I have mixed feelings about these changes. It's reasonable to use size_t for things that are the length of a structure or a string, but I don't think that the savings in code size will be significant either in space or time. There's also a lot of changes spread out across the source tree which could be better as separate commits.

AtariDreams commented:
> I have mixed feelings about these changes. It's reasonable to use size_t for things that are the length of a structure or a string, but I don't think that the savings in code size will be significant either in space or time. There's also a lot of changes spread out across the source tree which could be better as separate commits.

I think they would be best done at once, that's why, since they all have the same theme.

Neustradamus commented:
@AtariDreams: It is not possible to separate?

paulusmack commented:
I notice also that the signoff is a noreply.github.com address, which I don't consider adequate.

Neustradamus commented:
@AtariDreams: Have you seen the @paulusmack comment?
What is the situation with your signoff?

AtariDreams commented:
@Neustradamus Fixed!

paulusmack reviewed (commented):
In general it seems like there is a bunch of refactoring here plus gratuitous whitespace changes, in addition to the changes mentioned in the commit description (i.e. converting int to size_t). So, at the very least the commit description needs to be more comprehensive. If nothing else it should identify any potential bugs fixed here, though in fact I would prefer the bug fixes to be in separate commits.

I'm not convinced by the rationale that we need to do this to reduce the generated code size. As far as I can see it is only going to reduce the code size on 64-bit machines, not 32-bit machines, some of which might arguably be a bit short on memory. If you have a 64-bit processor you probably have a GB or more of memory, and saving at most a few hundred bytes of instruction memory is going to be in the noise.

Neustradamus commented:
@AtariDreams: Have you seen @paulusmack comments started:
- https://github.com/ppp-project/ppp/pull/415#pullrequestreview-1795621594

Note: The other PR has been merged :)

Neustradamus commented:
@AtariDreams: Have you looked comments from @paulusmack?

AtariDreams commented:
I'll take a look.

Neustradamus commented:
@AtariDreams: Have you looked?
It will be good to have a solution before 2.5.1 release...

paulusmack commented:
As far as I can see, there is no functional change here (or at least, none intended), so I am not going to wait for this.

AtariDreams commented on chat/chat.c:
```diff
@@ -1421,8 +1424,8 @@ int get_string(register char *string)
-		int rep_len = strlen (report_buffer);
-		if ((rep_len + 1) < sizeof(report_buffer)) {
+		size_t rep_len = strlen (report_buffer);
+		if (rep_len < sizeof(report_buffer) - 1) {
```
It just makes more sense.

AZero13 committed:
Prefer size_t over int where possible

This avoids needless truncation, which takes up multiple more instructions on some architectures, especially older ones.

Signed-off-by: Rose <gfunni234@gmail.com>

LumioseSil committed:
Prefer size_t over int where possible

This avoids needless truncation, which takes up multiple more instructions on some architectures, especially older ones.

Signed-off-by: Rose <gfunni234@gmail.com>

Neustradamus commented:
@paulusmack: Have you seen the @AtariDreams changes?

paulusmack commented on pppd/main.c:
```diff
@@ -1863,9 +1863,9 @@ update_script_environment(void)
 
     for (uep = userenv_list; uep != NULL; uep = uep->ue_next) {
-	int i;
+	size_t i;
```
For a case like this where the variable is a loop index rather than the size or length of something, I'd prefer plain long or unsigned long rather than size_t. Same comment applies in a few other places.

paulusmack commented on pppd/plugins/pppoatm/misc.c:
```diff
@@ -37,7 +37,7 @@ int __atmlib_fetch(const char **pos,...)
     }
     va_end(ap);
-    if (best > -1) (*pos) += best_len;
+    (*pos) += best_len;
```
This is a semantic change; previously the function would return -1 if the string at *pos didn't match any of the argument strings at all, whereas now it will return 0, which callers interpret as a match with the second argument. So now text2qos will parse things incorrectly.

This is an example of why I would have preferred this commit to be split into smaller pieces, for example, four separate commits for chat, pppd, and the pppoatm and pppoe plugins. If you had, I could merge the commits that are OK. As it is, I can't merge any of it.

paulusmack commented:
Overall, I would take everything except the pppoatm changes, which don't seem to me to be at all necessary, and in fact wrong, as I commented.

shahbaz606 reviewed (commented):
l

Neustradamus commented:
@AreaZR: Have you seen the latest @paulusmack comments?

Neustradamus commented:
@AZero13: Your PR will be not finished before next version? :/

Neustradamus commented:
@AZero13: Have you seen the latest @paulusmack comments?

AZero13 commented:
Yes I will address this now

Neustradamus commented:
@AZero13: I have sent you e-mails about your PR.

Can you look your PR?

Thanks in advance.

AZero13 commented:
Okay, okay okay I will. Yes I will do this

jkroonza commented on pppd/main.c:
```diff
@@ -2184,8 +2184,10 @@ ppp_script_setenv(char *var, char *value, int iskey)
 
     newstring = (char *) malloc(vl+1);
-    if (newstring == 0)
+    if (newstring == 0 || vl > INT_MAX) {
```
Please clarify.

jkroonza commented on pppd/options.c:
```diff
@@ -1673,6 +1675,8 @@ callfile(char **argv)
     }
 
     l = strlen(arg) + strlen(PPP_PATH_PEERFILES) + 1;
+    if (l > INT_MAX)
```
This does NOT look like a sensible way to protect against counter overflows.

jkroonza commented on pppd/options.c:
```diff
@@ -1873,9 +1877,8 @@ user_setenv(char **argv)
-	if (nlen == (eqp - arg) &&
-	    strncmp(arg, uep->ue_name, nlen) == 0)
+	size_t nlen = strlen(uep->ue_name);
+	if (nlen == (eqp - arg) && strncmp(arg, uep->ue_name, nlen) == 0)
```
This changes a line without changing it.  I don't object, but my rule of thumb is that unless you're touching the code in a line anyway, don't reformat it either.

jkroonza commented on pppd/tls.c:
```diff
@@ -186,12 +186,13 @@ static int tls_verify_callback(int ok, X509_STORE_CTX *ctx)
-            }
+
+            len1 = strlen(ptr1);
+            len2 = strlen(ptr2);
```
This is a logical breaking change.  There is no chance that this would work correctly.  Specifically tls_verify_method would need to be TLS_VERIFY_SUFFIX, at which point ptr2 won't be initialised by either of the previous two if statements.

strlen(NULL) will segfault.

jkroonza commented on pppd/tls.c:
```diff
@@ -186,12 +186,13 @@ static int tls_verify_callback(int ok, X509_STORE_CTX *ctx)
+                ptr2 += len2 - len1;
         }
 
         if (strcmp(ptr1, ptr2)) {
```
Unrelated to the change, but what happens if tls_verify_method doesn't match any of the if blocks above?  ptr2 == NULL => segfault.

jkroonza reviewed (approved):
I did recheck, and I'm no longer seeing functional changes, so from my perspective, this looks good now.

Neustradamus commented:
@AZero13: Never too late, thanks for your work and thanks to @paulusmack for merging!
